### ISSUE TITLE
malformed cgroup entry 1:net_cls:/

### ISSUE THREAD

**Issue by trb** [NVIDIA/libnvidia-container#150]
I installed nvidia-docker2 on Ubuntu21.10 which obviously isn't officially supported, and on my system it throws this error:

`docker: Error response from daemon: failed to create shim: OCI runtime create failed: container_linux.go:380: starting container process caused: process_linux.go:545: container init caused: Running hook #0:: error running hook: exit status 1, stdout: , stderr: nvidia-container-cli: container error: failed to get device cgroup root path: malformed cgroup entry: 1:net_cls:/: unknown.`

Which I think is thrown by this line, and I'm wondering if this line is actually required: https://github.com/NVIDIA/libnvidia-container/blob/ac25174795e992268820233967cdaf2c299166cf/src/nvcgo/internal/cgroup/v2.go#L85

more context:

```
func (c *cgroupv2) GetDeviceCGroupRootPath(procRootPath string, pid int) (string, error) {
(...)
  for scanner.Scan() {
    // Split each entry by ':'
    parts := strings.SplitN(scanner.Text(), ":", 3)
    if len(parts) != 3 {
	    return "", fmt.Errorf("malformed cgroup entry: %v", scanner.Text())
    }
    if parts[1] != "" {
	    return "", fmt.Errorf("malformed cgroup entry: %v", scanner.Text())
    }
    // Return the cgroup root from the 2nd element.
    return parts[2], nil
  }
(...)
}
```

I've got the following cgroup entries for most of my processes (I checked a few hundred, not all though):

```
1:net_cls:/
0::/
```

I.e. `1:net_cls:/` is the first entry in all process cgroups, but never the only one - there's always one with hierachy 0 and the second value == "".

I have no idea what I'm talking about or how cgroups work, but could it be that instead of erroring out on line 85 it would be possible to `continue` and only error out `if parts[1] != ''` was not true for _any_ entry?

**Comment by trb** [comment]
If anyone finds this issue because they ran into the same error message, I think it's due to upgrading the OS instead of doing a fresh install. Again, I have no idea how cgroup really works, but it's my impression that v1 controllers somehow can stick around even if the OS switched to v2 - _even_ if the hybrid mode is deactivated with the `splash systemd.unified_cgroup_hierarchy=1` boot option. You also need to add `cgroup_no_v1=all` to prevent any v1 controllers from being mounted. This fixed the issue with `nvidia-container-cli` for me, as the cgroup entries are now formed the way the library expects them to be.

So, in short, add this to /etc/default/grub:

`systemd.unified_cgroup_hierarchy=1 cgroup_no_v1=all`

My line:

`GRUB_CMDLINE_LINUX_DEFAULT="quiet splash systemd.unified_cgroup_hierarchy=1 cgroup_no_v1=all"`

And don't forget to run `sudo update-grub` after changing the file.

**Comment by klueska** [comment]
I think this line should probably be changed to a ‚continue‘. We would have previously verified that the „devices“ cgroup is governed by cgroupv2, but that doesn’t necessarily prohibit other cgroups from continuing to be governed by cgroupv1.

The current code here assumes there is only a single entry in this file and that it reports a cgroupv2 hierarchy (or at least that the first entry is such). I think it’s probably more correct to search through this file and find the cgroupv2 hierarchy than error out after not finding it in the first entry.

Will update in the next RC. Thanks for reporting.

**Comment by trb** [comment]
Awesome, I'm happy I could help a tiny bit :)

**Comment by klueska** [comment]
https://gitlab.com/nvidia/container-toolkit/libnvidia-container/-/merge_requests/123
